### ISSUE TITLE
Add CUDA implementation of Rotate function

### ISSUE THREAD

**Issue by ihhub** [ihhub/penguinV#267]
We have Rotate function implementation which locates in `src/image_function.h` and `src/image_function.cpp` files:
```cpp
void Rotate( const Image & in, double centerXIn, double centerYIn, Image & out, double centerXOut, double centerYOut, double angle );
```
We want to add CUDA implementation of the same function. Please put this implementation into `src/cuda/image_function_cuda.cuh` and `src/cuda/image_function_cuda.cu`. Use an existing code style of format to do it so it should pretty straightforward.

**Comment by MatthewMcGonagle** [comment]
I'm interested in taking a look at this, but I have a question about something inside `Image_Function::Rotate()`. It concerns the following code:
``` cpp
                        // we use bilinear approximation to find pixel intensity value
                        double coeffX = posX - x;
                        double coeffY = posY - y;

                        double sum = (*inX) * (1 - coeffX) * (1 - coeffY) +
                            (*inX + 1) * (coeffX) * (1 - coeffY) +
                            (*inX + rowSizeIn) * (1 - coeffX) * (coeffY)+
                            (*inX + rowSizeIn + 1) * (coeffX) * (coeffY)+0.5;

                        (*outX) = static_cast<uint8_t>(sum);
```

Its seems to me that `sum` is supposed to be an average of four pixels (although I'm not quite sure I understand the 0.5 offset). However, I think there is an error in the precedence of stuff like `*inX + 1`. My understanding is that the intention is that this is `*(inX + 1)`, but this is actually `(*inX) + 1`.

For example, the following minimal example
``` cpp
#include<iostream>

int main() {

    int nums[3] = {1, 10, 100};
    int *node = nums;

    std::cout << "*node = " << *node << std::endl
              << "*node + 1 = " << *node + 1 << std::endl
              << "*(node + 1) = " << *(node + 1) << std::endl;
    return 0;
}
```

gives as output of
```
*node = 1
*node + 1 = 2
*(node + 1) = 10
```

**Comment by MatthewMcGonagle** [comment]
Another question. 

It seems to me that quantities outside of the `for` loops , such as `cosAngle` or `inXPos`, should be calculated by `Image_Function_Cuda::Rotate()` before calling a kernel on the GPU. The computed quantities can just be passed into the kernel. So the threads of the kernel simply handle doing the `for` loops in parallel.

Does this seem good to you?

**Comment by ihhub** [comment]
Hi, @MatthewMcGonagle thanks for taking thia issue. Please take a look into coefficients for pixel intensities. Such variables are from 0 to 1 and combinations of them are from 0 to 1. The sum of all 4 coefficients is 1.

It's fine to pass precalculated values into GPU kernel (function) so you could do it.

Extra note: you might need to use float instead of double as this is faster without loosing accuracy.

**Comment by MatthewMcGonagle** [comment]
Hi @ihhub. Yes, I understand that the `sum` in `Image_Function::Rotate()` is a weighted average, but the problem is if it is taking the weighted average of the correct values? I believe the intent is to take a weighted average of the values at pixels `[inY][inX]`, `[inY][inX+1]`, `[inY+1][inX]`, and `[inY+1][inX+1]`, but the precedence of the dereferencing `*inX` happens before the addition. So `*inX + 1 == (*inX) + 1` instead of `*(inX + 1)`. 

Thanks for the tip on using `float`!

**Comment by ihhub** [comment]
I think you're right. Could you please debug the original code if you have a free time? I don't have an access to a PC until next Wednesday.

**Comment by MatthewMcGonagle** [comment]
@ihhub Yeah, sure I can take of the original code. Should I put the fix in a separate pull request, or should I just include it with the pull request for this issue?

**Comment by ihhub** [comment]
A separate pull request is preferable.

**Comment by MatthewMcGonagle** [comment]
@ihhub Okay, I've made a pull request for fixing the original code in `Image_Function::Rotate()`.

**Comment by MatthewMcGonagle** [comment]
@ihhub I've finished the CUDA version of `Rotate()`. I'll make a pull request.
